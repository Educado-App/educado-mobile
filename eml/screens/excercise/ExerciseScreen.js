import { StatusBar } from "expo-status-bar";
import React, { useEffect, useRef, useState } from "react";
import { Alert, StyleSheet, View, Text } from "react-native";
import LeaveButton from "../../components/exercise/LeaveButton";
import ExerciseButtons from "../../components/exercise/ExerciseButtons";
import { useNavigation, useRoute } from "@react-navigation/native";
import { Video } from "expo-av";
import * as StorageService from "../../services/StorageService";
import CustomProgressBar from "../../components/exercise/Progressbar";
import dummyExerciseData from './dummyExerciseData.json';



export default function ExerciseScreen() {
  const navigation = useNavigation();
  const route = useRoute();
  
  const [hasData, setHasData] = useState(false);
  const [signal, setSignal] = useState([]);
  const [exerciseData, setExerciseData] = useState({});

  /*async function getExercise() {
    const exercise = await StorageService.getNextExercise(sectionId);

    if (exercise !== null) {
      console.log(exercise);

      if (exercise === true) {
        navigation.navigate("SectionComplete", {
          courseId: courseId,
          sectionId: sectionId,
        });
      } else {
        setExerciseData(exercise);
      }
    } else {
      navigation.navigate("ErrorScreen");
    }
  }*/

  async function getExercise() {
    // Instead of calling StorageService.getNextExercise, use the dummy data
    const exercise = dummyExerciseData;
  
    if (exercise !== null) {
      console.log(exercise);
  
     /* const courseId = exercise.courseId || "defaultCourseId";
      const sectionId = exercise.sectionId || "defaultSectionId";*/
  
      if (exercise === true) {
       /* navigation.navigate("SectionComplete", {
          courseId: courseId,
          sectionId: sectionId,
        });*/
      } else {
        setExerciseData(exercise);
      }
    } else {
      navigation.navigate("ErrorScreen");
    }
  }
  
  

  /*async function isSectionComplete(courseId, sectionId) {
    const course = await StorageService.getCourseById(courseId);
    const sections = course.sections;

    for (let i = 0; i < sections.length; i++) {
      console.log("Section is complete: ", sections[i].isComplete);
      if (sections[i].id === sectionId) {
        return sections[i].isComplete;
      }
    }
  }*/

  useEffect(() => {
    getExercise().then(() => {
      setHasData(true);
    });
  }, [route.params]);

  const video = useRef(0);

  useEffect(() => {
    if (video !== undefined) {
      if (signal === 0) {
        try {
          video.current.pauseAsync();
        } catch (e) {
          console.log("Something went wrong :" + e);
        }
      }
      setSignal(1);
    }
  }, [signal]);

  return (
    <View style={styles.container} className="bg-babyBlue">
      <View style={{ flex: 1 }}>
        <View className="pt-8 flex-row w-screen">
          <View className="">
            <LeaveButton
              //navigationPlace={"Course"}
              //courseId={courseId}
            ></LeaveButton>
          </View>
          <View
            style={{
              left: "80%",
              justifyContent: "center",
              alignItems: "center",
              flexDirection: "column",
            }}
          ></View>
        </View>
        <View className="items-center">
          <CustomProgressBar progress={0.5 / 1}></CustomProgressBar>
        </View>
      </View>
      <View style={{ flex: 2, width: "100%" }}>
        {exerciseData === undefined ? (
          // No Data
          <Text> Sem dados</Text>
        ) : (
          <Video
            source={{ uri: exerciseData.content }}
            rate={1.0}
            volume={1.0}
            isMuted={false}
            resizeMode="cover"
            shouldPlay
            useNativeControls
            ref={video}
            style={styles.backgroundVideo}
          />
        )}
      </View>
      <View style={{ flex: 3 }}>
        {exerciseData === undefined ? (
          //No data
          <Text> Sem dados</Text>
        ) : (
          <ExerciseButtons
            answers={exerciseData.answers}
            exerciseId={exerciseData.id}
            courseId={exerciseData.courseId}
            sectionId={exerciseData.sectionId}
            setSignal={setSignal}
            hasData={hasData}
          ></ExerciseButtons>
        )}
      </View>
      <StatusBar style="auto" />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
  },
  row: {
    flexDirection: "row",
  },
  textStyle: {
    zIndex: 100,
    fontSize: 100,
  },
  buttonShadow: {
    shadowOpacity: 0.4,
    shadowRadius: 10,
    shadowOffset: { width: 0, height: 0 },
  },
  backgroundVideo: {
    height: "100%",
  },
});
