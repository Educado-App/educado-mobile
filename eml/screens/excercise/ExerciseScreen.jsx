import React, { useState } from 'react';
import { ScrollView, View, TouchableOpacity } from 'react-native';
import Text from '../../components/general/Text';
import { RadioButton } from 'react-native-paper';
import ExerciseInfo from '../../components/exercise/ExerciseInfo';
import { Icon } from '@rneui/themed';
import { SafeAreaView } from 'react-native-safe-area-context';
import PopUp from '../../components/gamification/PopUp';
import { StatusBar } from 'expo-status-bar';
import PropTypes from 'prop-types';
import { completeComponent, handleLastComponent } from '../../services/utilityFunctions';
import { useNavigation } from '@react-navigation/native';

/* 
Description: 	This screen is displayed when the student is doing an exercise.
				It displays the question and the answers, and the student can select one answer.
				When the student presses the confirm button, the answer is checked and the student is given feedback.
				When the student presses the continue button, the next component is displayed.
				The student can only continue if an answer is selected.
				The student is given 10 points when the answer is correct in the first try, 
				otherwise the student gets 5 points when the answer is correct.
				The student gets 0 points when the answer is incorrect or they have completed the exercise before.
Dependencies: 	CompSwipeScreen, the screen which contains all the components in the section
Props:			- exerciseObject: The exercise object, which contains the question and the answers
				- sectionObject: The section object, which contains the section title	
				- courseObject: The course object, which contains the course title
				- onContinue: A function that is called when the student presses the continue button, 
				when the exercise is completed and it is the last component in the section, the student is taken to the section complete screen
*/

export default function ExerciseScreen({ exerciseObject, sectionObject, courseObject, onContinue }) {
	const tailwindConfig = require('../../tailwind.config.js');
	const projectColors = tailwindConfig.theme.colors;
	const navigation = useNavigation();

	const [selectedAnswer, setSelectedAnswer] = useState(null);
	const [buttonClassName, setButtonClassName] = useState('');
	const [showFeedback, setShowFeedback] = useState(false);
	const [buttonText, setButtonText] = useState('Confirmar Resposta'); 
	const [isPopUpVisible, setIsPopUpVisible] = useState(false); 
	const [isCorrectAnswer, setIsCorrectAnswer] = useState(false);
	const [points, setPoints] = useState(10);

	const handleAnswerSelect = (answerIndex) => {
		setSelectedAnswer(answerIndex);
	};
  
	async function handleReviewAnswer(selectedAnswer) {
		const continueText = 'Continuar';

		setIsCorrectAnswer(selectedAnswer);

		setButtonClassName(
			`bg-project${selectedAnswer ? 'Green' : 'Red'}`
		);

		setShowFeedback(true);
		setButtonText(continueText);
		if (buttonText !== continueText) {
			const obj = await completeComponent(exerciseObject, courseObject.courseId, selectedAnswer);
			setPoints(obj.points);
			setIsPopUpVisible(true);
		} else {
			setIsPopUpVisible(false);
			if (onContinue()) {
				handleLastComponent(exerciseObject, courseObject, navigation);
			}
		}
	}
  

	return (
		<SafeAreaView className="h-full bg-secondary">

			<View className='items-center'>
				<Text testID='exerciseQuestion'
					className='pt-20 pb-10 text-center text-body font-sans-bold text-projectBlack w-11/12'>
					{exerciseObject.question}
				</Text>

				<View className={`${buttonClassName} items-center justify-center h-96 w-full`}>
					<ScrollView className="py-2">
						{/* Map through the answers and render each one */}
						{exerciseObject.answers.map((answer, index) => (
							<View
								key={index}
								className='flex-row w-96 pb-6 pl-2'
							>
								<View>
									<RadioButton.Android
										disabled={showFeedback}
										value={index}
										status={
											selectedAnswer === index ? 'checked' : 'unchecked'
										}
										onPress={() => handleAnswerSelect(index)}
										color={projectColors.primary}
										uncheckedColor={projectColors.primary}
									/>
								</View>

								<View>
									<TouchableOpacity onPress={() => handleAnswerSelect(index)} disabled={showFeedback}>
										<Text className='pt-2 pb-1 w-72 font-montserrat text-body text-projectBlack'>{answer.text}</Text>
									</TouchableOpacity>

									{showFeedback ? (
										<View className={`flex-row pb-2 w-fit rounded-medium ${answer.correct ? 'bg-projectGreen' : 'bg-projectRed'}`}>
											<View className='pl-2 pt-1'>
												<View className='pt-1.5'>
													{answer.correct === true ? (
														<Icon
															size={10}
															name='check'
															type='material'
															color={projectColors.success}
														/>
													) : (
														<Icon
															size={10}
															name='close'
															type='material'
															color={projectColors.error}
														/>
													)}
												</View>
											</View>
											<Text className={`w-72 pl-1 pt-2 pr-2 text-caption-medium ${answer.correct ? 'text-success' : 'text-error'}`}>{answer.feedback}</Text>
										</View>
									) : null}
								</View>
							</View>
						))}
					</ScrollView>
				</View>
				<View className='px-6 pt-10 w-screen'>
					<TouchableOpacity
						disabled={selectedAnswer === null ? true : false}
						className={`${selectedAnswer !== null ? 'opacity-100' : 'opacity-30'} bg-primary px-10 py-4 rounded-medium`}
						onPress={() => handleReviewAnswer(exerciseObject.answers[selectedAnswer].correct)}
					>
						<Text className='text-center font-sans-bold text-body text-projectWhite'>{buttonText}</Text>
					</TouchableOpacity>
				</View>
			</View>
    

			{isPopUpVisible ? (
				<PopUp pointAmount={points} isCorrectAnswer={isCorrectAnswer} />
			) : null}

			{<ExerciseInfo courseTitle={courseObject.title} sectionTitle={sectionObject.title} />}
			<StatusBar style='auto' />
		</SafeAreaView>
	);
}

ExerciseScreen.propTypes = {
	exerciseObject: PropTypes.object,
	sectionObject: PropTypes.object,
	courseObject: PropTypes.object,
	onContinue: PropTypes.func,
};

/*
async function getExercise() {
  const exercise = await StorageService.getNextExercise(sectionId);

  if (exercise !== null) {
    console.log(exercise);

    if (exercise === true) {
      navigation.navigate("SectionComplete", {
        courseId: courseId,
        sectionId: sectionId,
      });
    } else {
      setExerciseData(exercise);
    }
  } else {
    navigation.navigate("ErrorScreen");
  }
}*/

/*async function isSectionComplete(courseId, sectionId) {
const course = await StorageService.getCourseById(courseId);
const sections = course.sections;

for (let i = 0; i < sections.length; i++) {
  console.log("Section is complete: ", sections[i].isComplete);
  if (sections[i].id === sectionId) {
    return sections[i].isComplete;
  }
}
}

async function getExercise() {
  // Instead of calling StorageService.getNextExercise, use the dummy data
  const exercise = dummyExerciseData;

  if (exercise !== null) {
    const courseId = exercise.courseId || "defaultCourseId";
    const sectionId = exercise.sectionId || "defaultSectionId";

    if (exercise === true) {
        navigation.navigate("SectionComplete", {
        courseId: courseId,
        sectionId: sectionId,
      });
    } else {
      setExerciseData(exercise);
    }
  } else {
    navigation.navigate("ErrorScreen");
  }
}
*/

