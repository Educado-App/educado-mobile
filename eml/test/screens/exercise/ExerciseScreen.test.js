import React from "react";
import renderer from "react-test-renderer";
import ExerciseScreen from "../../../screens/excercise/ExerciseScreen";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { act } from "react-test-renderer"; // Import act for async testing

let navigated = false;

jest.mock("@react-navigation/native", () => ({
  useNavigation: () => ({
    navigate: jest.fn(() => {
      navigated = true;
    }),
  }),
  useRoute: () => ({
    params: {},
  }),
}));

// Mock your asynchronous functions here
jest.mock("../../../api/api", () => ({
  getExercise: jest.fn().mockResolvedValue("65181a4f4c78b45368126ed7"),
  getSection: jest.fn().mockResolvedValue({
    exercises: ["65181a4f4c78b45368126ed7", "6527e20b5897d93ea8914010"],
    components: [],
    _id: "6526528fc808771dd861fbec",
    parentCourse: "651e987f8999e62500457097",
    title: "Test X",
    description: "Hello",
    dateCreated: "2023-10-11T07:45:19.135Z",
    dateUpdated: "2023-10-11T07:45:19.135Z",
    __v: 37,
    lectures: [],
    createdAt: "2023-10-19T10:54:26.299Z",
    modifiedAt: "2023-10-19T10:54:26.299Z",
  }),
  getCourse: jest.fn().mockResolvedValue({
    rating: 2,
    status: "draft",
    numOfSubscriptions: 0,
    sections: [
      "651e99608999e625004570a8",
      "6526528fc808771dd861fbec",
      "6527f7da1877373c3c653dc8",
    ],
    _id: "651e987f8999e62500457097",
    title: "VirtualTurtor Group",
    category: "FinanÃ§as pessoais",
    level: "Iniciante",
    description: "Silke var her",
    dateCreated: "2023-10-05T11:05:35.251Z",
    dateUpdated: "2023-10-05T11:05:35.251Z",
    __v: 11,
    published: false,
    estimatedHours: 1,
  }),
}));

describe("ExerciseScreen", () => {
  beforeEach(() => {
    navigated = false;
    AsyncStorage.clear();
  });

  it("renders ExerciseScreen correctly", async () => {
    let tree;

    // Wrap the rendering inside act for async testing
    await act(async () => {
      tree = renderer.create(<ExerciseScreen />);
    });

    // Use toTree to get the rendered tree
    expect(tree.toTree()).toMatchSnapshot();
  });
});
