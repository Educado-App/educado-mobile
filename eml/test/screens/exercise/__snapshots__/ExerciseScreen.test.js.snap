// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseScreen renders ExerciseScreen correctly 1`] = `null`;
